scanner/nodeFramework.go

```diff
@@ -0,0 +1,175 @@
+		return nil, nil
+	} else {
+		nodeVersion := strings.TrimSpace(string(out))
+		if nodeVersion[:1] == "v" {
+			nodeVersion = nodeVersion[1:]
+		}
+		if nodeVersion < "16" {
+			return nil, nil
+		}
+	}
+
+	srcInfo := &SourceInfo{
+		Family:     "NodeJS",
+		SkipDeploy: true,
+		Callback:   NodeFrameworkCallback,
+	}
+
+	return srcInfo, nil
+}
+
+func NodeFrameworkCallback(srcInfo *SourceInfo, options map[string]bool) error {
+	// generate Dockerfile if it doesn't already exist
+	_, err := os.Stat("Dockerfile")
+	if errors.Is(err, fs.ErrNotExist) {
+		var args []string
+
+		_, err = os.Stat("node_modules")
+		if errors.Is(err, fs.ErrNotExist) {
+			// no existing node_modules directory: run package directly
+			args = []string{"npx", "--yes", "@flydotio/dockerfile@latest"}
+		} else {
+			// build command to install package using preferred package manager
+			args = []string{"npm", "install", "@flydotio/dockerfile", "--save-dev"}
+
+			_, err = os.Stat("yarn.lock")
+			if !errors.Is(err, fs.ErrNotExist) {
+				args = []string{"yarn", "add", "@flydotio/dockerfile", "--dev"}
+			}
+
+			_, err = os.Stat("pnpm-lock.yaml")
+			if !errors.Is(err, fs.ErrNotExist) {
+				args = []string{"pnpm", "add", "-D", "@flydotio/dockerfile"}
+			}
+		}
+
+		// check first to see if the package is already installed
+		installed := false
+
+		deps, ok := packageJson["dependencies"].(map[string]interface{})
+		if ok && deps["@flydotio/dockerfile"] != nil {
+			installed = true
+		}
+
+		deps, ok = packageJson["devDependencies"].(map[string]interface{})
+		if ok && deps["@flydotio/dockerfile"] != nil {
+			installed = true
+		}
+
+		// install/run command
+		if !installed || args[0] == "npx" {
+			cmd := exec.Command(args[0], args[1:]...)
+			cmd.Stdin = nil
+			cmd.Stdout = os.Stdout
+			cmd.Stderr = os.Stderr
+
+			if err := cmd.Run(); err != nil {
+				return errors.Wrap(err, "Failed to install @flydotio/dockerfile")
+			}
+		}
+
+		// run the package if we haven't already
+		if args[0] != "npx" {
+			cmd := exec.Command("npx", "dockerfile")
+			cmd.Stdin = nil
+			cmd.Stdout = os.Stdout
+			cmd.Stderr = os.Stderr
+
+			if err := cmd.Run(); err != nil {
+				return errors.Wrap(err, "Failed to generate Dockerfile")
+			}
+		}
+	}
+
+	// read dockerfile
+	dockerfile, err := os.ReadFile("Dockerfile")
+	if err != nil {
+		return errors.Wrap(err, "Dockerfile not found")
+	}
+
+	// extract family
+	family := "NodeJS"
+	re := regexp.MustCompile(`(?m)^LABEL\s+fly_launch_runtime="(?P<family>.+?)"`)
+	m := re.FindStringSubmatch(string(dockerfile))
+
+	for i, name := range re.SubexpNames() {
+		if len(m) > 0 && name == "family" {
+			family = m[i]
+		}
+	}
+	srcInfo.Family = family
+
+	// extract port
+	port := 3000
+	re = regexp.MustCompile(`(?m)^EXPOSE\s+(?P<port>\d+)`)
+	m = re.FindStringSubmatch(string(dockerfile))
+
+	for i, name := range re.SubexpNames() {
+		if len(m) > 0 && name == "port" {
+			port, err = strconv.Atoi(m[i])
+			if err != nil {
+				panic(err)
```

Comment:
How about returning this err instead of panic?

Reply:
My reasoning is as follows: I used a regular expression.  At this point, the match should be all digits.  If Atoi has a problem with it, there is nothing reasonable a user can do to address the problem, there clearly is a software bug.  Full disclosure: I copied this code from elsewhere, but as far as I can tell, the reasoning is sound.

Reply:
I see. That makes sense.